[그리디] 염지환 자동차 경주 1, 2단계 제출

JihwanYeom opened the pull request:
안녕하세요! 그리디에서 첫 과제를 제출하게 된 염지환이라고 합니다.

처음에는 과제 요구 사항을 보고 매우 간단하다고 생각했으나 직접 코드를 작성할수록 제가
처음 코딩을 배운 때로 돌아간 것 같다는 느낌을 받았습니다.

프로그래밍 요구 사항에 맞추어 코드를 작성하려고 하니 생각해야 할 부분이 많았습니다.
특히 단일 책임 원칙을 지키기 위해 최대한 함수의 기능을 간결하게 구현하였는데 이 때 어떤 규모의 기능까지
한 단위로 묶어야 할 지 고민하는데 시간을 많이 쏟았습니다. 너무 함수를 잘게 나누면 오히려 코드를 파악하기
힘들어 지지는 않을까 라는 생각도 했던 것 같습니다.

또한 필드와 메서드 이름 역시 고민을 많이 했습니다. 제가 기준으로 삼은 목표는 호출된 메서드의 구현 내용을
확인하지 않고도 현재 메서드가  어떻게 동작하는지 이해할 수 있는 코드를 작성하는 것이었습니다. 저는 어느 정도
목표를 달성했다고 생각하나 부족한 부분이 있다면, 혹은 더 좋은 목표 기준이 있다면 지적해주시면 감사하겠습니다.

테스트 코드의 존재도 이번 과제를 통해 알게 되어 열심히 찾아보고 작성했으나 좋은 테스트 코드인지는
아직 의문이 많습니다. 
테스트 코드를 작성하기 위해 클래스 구조도 많이 고쳤습니다. 확실히 테스트 코드 작성을 염두하고
클래스를 수정하니 훨씬 객체지향적으로 좋은 코드라는 걸 직관적으로 느낄 수 있었습니다.

미처 신경 쓰지 못한 부분 많이 지적해주시면 감사하겠습니다!

JihwanYeom committed:
* feat : 1단계 구현

JihwanYeom committed:
* test : Car 객체 테스트 코드 작성

JihwanYeom committed:
* feat : 2단계 구현

JihwanYeom committed:
* test : makeWinnerList() 테스트코드 작성

JihwanYeom committed:
* docs : README 작성

TaeyeonRoyce commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+---
+
+## 📋프로젝트 개요
+이 프로젝트는 N개의 자동차가 참여하여 경주하는 간단한 레이싱 게임입니다   
+참여한 자동차들은 게임 규칙에 따라 각 라운드마다 1칸 전진, 또는 정지할 수 있으며   
+모든 라운드가 끝난 뒤 가장 많이 이동한 자동차들이 우승하게 됩니다.
```
프로젝트에 대한 개요가 작성 되어 있어 파악하기 좋네요! 👍

TaeyeonRoyce commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+* 🏆 모든 라운드가 끝나고 우승한 자동차를 확인하면 됩니다.
+
+## 🔧프로젝트 설계
+프로젝트는 크게 Car, Game 두개의 클래스로 작성되었습니다.
+### 🚗Car 클래스
+* Car 클래스는 필드로 이름과 이동 거리를 가지고 있습니다.
+* 생성자는 이름을 매개변수로 받아 초기화하고 이동거리를 0으로 초기화합니다.
+* move() 메서드는 매개변수로 받은 값이 4 이상일 경우 이동거리를 1 증가시킵니다.
+
+### 🎮Game 클래스
+* 참가하는 차들과 우승한 차를 저장하기 위해 두 개의 List<Car\> 필드를 가지고 있습니다.
+* playGame() 메서드는 차 갯수와 라운드 수를 매개변수로 받아 게임 초기화, 게임 진행, 우승자 선정을 수행합니다.
+* initGame() 메서드는 playGame()메서드에서 가장 먼저 호출되어 차 갯수를 매개변수로 받아 갯수만큼 Car 객체를 생성하여 리스트에 저장합니다. 현 단계에서는 차 이름이 순차적으로 자동생성됩니다.( ex)Car_1, Car_2, ... )
+* playRound() 메서드는 라운드 수만큼 실행되며 게임에 참여한 모든 Car클래스의 move()메서드를 호출합니다.
+* makeWinnerList() 메서드는 모든 라운드가 끝난 후 가장 먼 거리를 이동한 차들을 우승자 리스트에 저장합니다.
```
프로젝트 내부 코드에 대한 설명이 자세하게 나와있군요!
친절한 설명은 좋지만, 이렇게 코드에 의존적이면 관리 대상이 되기도 합니다.
예를 들어, 코드에서 `playGame()`이라는 메서드의 이름이 수정 된다면 코드 수정과 함께 명세서도 수정해야 하죠.

코드가 아닌 미션과 관련한 요구사항(로직)을 작성해도 좋아보여요! 코드 수정은 빈번하지만 요구사항은 비교적으로 덜 변하니깐요.
추상적으로 작성 할수록 변경 영역은 줄어들어요. 그렇다고 너무 추상적이면 의미를 담아내기 어렵겠죠?
ex)
`move() 메서드는 매개변수로 받은 값이 4 이상일 경우 이동거리를 1 증가시킵니다.`
=> `자동차는 매개변수로 받은 값이 4 이상인 경우 이동거리를 1 증가시킵니다.`
=> `자동차는 랜덤한 값에 따라 이동거리가 증가합니다`
=> `자동차는 특정 조건에 따라 앞으로 나아갑니다`

코드에 의존적이지 않도록 작성하는게 좋을지, 어디까지 추상적으로 적으면 좋을지, 이 설계/명세를 누가 읽는지도 고민해보시죠!

TaeyeonRoyce commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,24 @@
+public class Car {
+
+    private String name;
```
`final`이 가능한 필드라면 안전성을 위해 붙혀주세요!

JihwanYeom replied:
아직 다른 언어의 const에 더 익숙해서 입력을 받는 변수에 대해 final을 쓰는 것에 조심스러웠는데
찾아보니 final 클래스 필드는 생성자에서 초기화해주는 것은 허용이 되는군요! 미처 잊고 넘어갔던
부분인데 지적해주셔서 감사합니다!

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+
+public class Game {
+
+    List<Car> carList = new ArrayList<Car>();
+    List<Car> winnerList = new ArrayList<Car>();
```
```suggestion
    List<Car> carList = new ArrayList<>();
    List<Car> winnerList = new ArrayList<>();
```

TaeyeonRoyce commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public void move(int randomNumber) {
+        if (randomNumber >= 4) {
```
4는 어떤 의미를 가질까요?
상수로 분리해서 의미를 부여한다면 더 읽기 좋을 것 같아요!

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+
+    public void initGame(int carNum){
+        for (int i = 0 ; i < carNum ; i++)
+            carList.add(new Car("Car_"+i+""));
```
불필요한 부분은 잘 확인하시고 지워주세요!
```suggestion
            carList.add(new Car("Car_"+i));
```

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    public int getRandom(){
+        return (int)(Math.random()*10);
+    }
```
랜덤 값 생성을 `Math`를 통해 하셨군요!
Java에서 랜덤값을 생성하는 방식이 여러가지 있는데요.

시드, 멀티스레드 성능, 난수 보안 등 여러 요인을 고려해서 다양한 선택을 할 수 있어요!
`Random`, `SecureRandom`, `ThreadLocalRandom`, `SplittableRandom` 등등
현재 미션에서는 `Math.random()`도 충분해 보이지만, 여러가지 방식들도 학습 해보시면 도움 되실거에요!

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void makeWinnerList() {
+        Car winner = carList.get(0);
```
사용하지 않는 라인은 항상 잘 체크하고 지워주세요!

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+        }
+        return maxDistance;
+    }
+}
```
EOF 때문에 경고가 발생했어요.
https://jojoldu.tistory.com/673
intellij 에서 자동으로 추가하도록 설정 할 수 있어요

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private void addWinner(Car car, int maxDistance) {
+        if (car.getDistance() == maxDistance) {
+            winnerList.add(car);
+        }
```
car의 position 값을 꺼내 값을 비교하는 것보단, car에서 해결하면 어떨까요?
```suggestion
        if (car.isSamePosition(maxDistance)) {
            winnerList.add(car);
        }
```
`Car`에게 메세지를 전달하는 방식으로 개선 할 수 있을 것 같아요!

TaeyeonRoyce commented on src/main/java/Game.java:
```diff
@@ -0,0 +1,54 @@
+    private int getMaxDistance() {
+
+        int maxDistance  = 0;
+        for (int i = 0; i < carList.size(); i++) {
+            maxDistance = Math.max(maxDistance, carList.get(i).getDistance());
+        }
```
인덱스가 필요 없다면, 아래와 같이 enhanced for-loop 를 사용하는게 어떨까요!
```suggestion
        for (Car car : carList) {
            maxDistance = Math.max(maxDistance, car.getDistance());
        }
```

TaeyeonRoyce commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,24 @@
+    @Test
+    void testCarName(){
+        Car car = new Car("KIA");
+        assertEquals("KIA", car.getName());
```
Junit에서 제공하는 assertion을 사용했군요!
이것도 물론 좋지만, 검증 단계에서는 assertj를 쓰시는걸 강력히 권장드려요.
https://kkoon9.tistory.com/135
가독성 뿐만 아니라 검증에 필요한 다양한 기능들을 정말 많이 제공합니다.

한번 사용해보시죠!

TaeyeonRoyce reviewed (changes requested):
안녕하세요 지환님! 메인테이너 겸 리뷰어로 활동중인 원태연입니다. 반가워요🙌
미션 하시느라 고생 많으셨습니다, 지환님!
충분한 고민이 묻어있는 코드 같았어요.

남겨주신 고민들도 읽어봤는데요. 메서드가 잘 분리되어 있는 것 같아 쉽게 읽혔어요.
말씀대로 메서드를 많이 나누다 보면 코드가 길어지고 오히려 가독성이 저해 될 수 있지만, 지환님이 일관 된 기준을 가지고(단일 책인 원칙) 분리를 한 덕분인지 코드를 이해하기 좋았어요.

특히, `호출된 메서드의 구현 내용을
확인하지 않고도 현재 메서드가 어떻게 동작하는지 이해할 수 있는 코드`를 고민하신 덕분에 메서드도 이름을 통해 파악하기 용이했습니다.
저도 동일한 기준으로 작성해요. 내가 작성한 메서드 시그니처가 메서드의 역할을 충분히 설명하는지, 이름 외의 다른 역할을 하는건 아닌지 항상 점검합니다. 좋은 고민이에요!

추가로, 제출하기 전에 [`Reformat`](https://eroul-ri.tistory.com/26) 한번씩 해주세요~!
미션하시느라 고생 많으셨고, 몇가지 코멘트 남겼으니 확인 부탁드려요!

JihwanYeom committed:
* docs : README 수정

추상화를 통한 변경 영역 축소

JihwanYeom committed:
* fix : Car, Game 클래스 리뷰 1차 반영

공통
Reformat을 통한 서식 통일

Car 클래스
name 필드 final 지정
가독성 증가를 위한 상수 추가
isSamePosition()메서드 추가

Game 클래스
리스트 생성 시 <>연산자 사용
미사용 라인 삭제
난수 생성 방식 변경
마지막 줄 개행 추가
enhanced for loop 사용

JihwanYeom committed:
* test : 테스트 코드 리뷰 1차 반영

가독성 향상을 위한 상수 필드 추가
assertj를 이용한 테스트로 변경

TaeyeonRoyce commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,29 @@
+
+public class CarTest {
+
+    private static final int MOVE_FORWARD = 4;
+    private static final int NOT_MOVE = 3;
```
직관적이고 좋은 이름이네요👍

TaeyeonRoyce commented on src/test/java/GameTest.java:
```diff
@@ -0,0 +1,24 @@
+        game.carList.get(2).move(MOVE_FORWARD);
+
+        game.makeWinnerList();
+        assertThat(game.winnerList).containsOnly(game.carList.get(0), game.carList.get(2));
```
잘 활용하시는군요👍

TaeyeonRoyce commented on README.md:
```diff
@@ -0,0 +1,44 @@
+
+
+* 라운드 진행 메서드는 라운드 수만큼 실행되며 게임에 참여한 모든 차의 이동 메서드를 호출합니다.   
+이때 각 자동차마다 랜덤 변수 생성기를 통해 0부터 10까지의 숫자 중 하나를 매개변수로 줍니다
```
👀
```suggestion
이때 각 자동차마다 랜덤 변수 생성기를 통해 0부터 9까지의 숫자 중 하나를 매개변수로 줍니다
```

TaeyeonRoyce commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,30 @@
+        }
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public int getDistance() {
+        return distance;
+    }
+
+    public boolean isSamePosition(int distance) {
+        return distance == this.distance;
+    }
```
메서드의 순서도 규칙을 지킨다면 가독성이 올라간답니다!
https://nooblette.tistory.com/entry/%EB%A9%94%EC%84%9C%EB%93%9C-%EB%B0%8F-%ED%95%84%EB%93%9C-%EC%84%A0%EC%96%B8-%EC%88%9C%EC%84%9C-%EC%BB%A8%EB%B2%A4%EC%85%98-%EC%9E%90%EB%B0%94
참고하세요~

TaeyeonRoyce reviewed (approved):
고생하셨어요, 지환님!
리뷰도 충분히 소화하시고 잘 반영해주셨네요!

다음 단계도 화이팅입니다!!
